cpp/harvest/nacjd_data_publication.cc

```diff
@@ -0,0 +1,276 @@
+/** \brief Utility for harvesting metadata from NACJD (ICPSR) - Publication.
+ *  \author Steven Lolong (steven.lolong@uni-tuebingen.de)
+ *
+ *  \copyright 2024 Tübingen University Library.  All rights reserved.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <fstream>
+#include <iostream>
+#include <nlohmann/json.hpp>
+#include "MARC.h"
+#include "StringUtil.h"
+
+
+using json = nlohmann::json;
+
+namespace {
+
+[[noreturn]] void Usage() {
+    ::Usage(
+        "[mode] [mode_params]\n"
+        "construct input_file output_file\n"
+        "\t- input_file: source of data in JSON format (taken from NACJD website).\n"
+        "\t- output_file: will contain all icpsr records as MARC21.\n"
+        "\n");
+}
+
+struct NacjdDoc {
+    std::string ref_id_, // REF_ID
+        title_,          // TITLE
+        issn_,           // ISSN
+        volume_,         // VOLUME
+        journal_,        // JOURNAL
+        i_number_,       // INUMBER
+        year_pub_,       // YEAR_PUB
+        place_pub_,      // PLACE_PUB
+        sec_title_,      // SEC_TITLE
+        ris_type_,       // RIS_TYPE, ex.: RPRT, JOUR, etc.
+        page_start_,     // PAGE_START
+        page_end_,       // PAGE_END
+        doi_,            // DOI
+        url_,            // URL
+        url_pdf_,        // URL_PDF
+        url_abs_,        // URL_ABS
+        type_work_,      // TYPE_WORK
+        publisher_;      // PUBLISHER
+
+    std::vector<std::string> author_split_, // AUTHORS_SPLIT
+        study_title_;                       // STUDYTITLE
+};
+void WriteMarc(const std::string &marc_path, const std::vector<NacjdDoc> &nacjd_docs) {
+    auto marc_file(MARC::Writer::Factory(marc_path));
+    MARC::Writer * const marc_writer(marc_file.get());
+
+    for (auto nacjd_doc : nacjd_docs) {
+        if (nacjd_doc.ref_id_.empty())
+            continue;
+
+        // construct author_info
+        std::vector<MARC::Subfield> authors;
+        if (not nacjd_doc.author_split_.empty()) {
+            for (auto author_ : nacjd_doc.author_split_) {
+                authors.emplace_back(MARC::Subfield('a', author_));
+            }
+        }
+        // construct publishing information
```

Comment:
The target field probably depends on the record type, e.g. for articles probably 773 (and for our Defaultlieferweg still 936).